src/features/form/hooks/useForm.ts

```diff
@@ -0,0 +1,121 @@
+import type { ChangeEvent, FocusEvent, FormEvent } from 'react'
+import { useCallback, useMemo, useState } from 'react'
+
+import { mapObject } from '~/utils/mapObject'
+
+import type { Validators } from '../types'
+
+type FormConfig<TFormData extends object> = {
+  initialValues?: Partial<TFormData>
+  validators?: Validators<TFormData>
+}
+
+const defaults = {
+  initialValues: {},
+  validators: {},
+}
+
+/**
+ * Execute a map of validators on values.
+ */
+const validate = <TFormData extends object>(
+  values: Partial<TFormData>,
+  validators: Validators<TFormData>
+) =>
+  mapObject(validators, (fn, field) =>
+    fn ? fn(values[field] ?? undefined) ?? null : null
+  )
+
+/**
+ * Controls login form state.
+ */
+const useForm = <TFormData extends object>(
```

Comment:
I was just curious, if i use Record<string, any> instead of object, that would lead to stronger typing right?

Reply:
Not really, I would say. `object` is more strict than `Record<string, any>`. We want the `TFormData` to be either provided or inferable from the `initialValues`. The `object` here helps with creating awareness when I _cannot_ infer nor have the type passed in:

**With `TFormData extends object` (warns):**
![Screen Shot 2022-06-16 at 17 21 05](https://user-images.githubusercontent.com/2688512/174104233-b653cf9b-2e15-4249-a5b7-c2aa1f46db5f.png)

**With `TFormData extends Record<string, any>` (does not warn):**
![Screen Shot 2022-06-16 at 17 21 45](https://user-images.githubusercontent.com/2688512/174104331-68a2e07c-3049-426e-89bf-ec212091b836.png)

Reply:
Not really, I would say. `object` is more strict than `Record<string, any>`. We want the `TFormData` to be either provided or inferable from the `initialValues`. The `object` here helps with creating awareness when I _cannot_ infer nor have the type passed in:

**With `TFormData extends object` (warns):**
![Screen Shot 2022-06-16 at 17 21 05](https://user-images.githubusercontent.com/2688512/174104233-b653cf9b-2e15-4249-a5b7-c2aa1f46db5f.png)

**With `TFormData extends Record<string, any>` (does not warn):**
![Screen Shot 2022-06-16 at 17 21 45](https://user-images.githubusercontent.com/2688512/174104331-68a2e07c-3049-426e-89bf-ec212091b836.png)

Reply:
Aaah, i see. Thank you, that make sense 💯  I saw the one with the Record in some example, now i see the difference.